File: LinkedLists/DoublyLinkedList.cpp
#include <bits/stdc++.h>
using namespace std;

template <class T>
class DoublyLinkedList
{
    class Node
    {
    private:
        T value;
        Node *next;
        Node *prev;

    public:
        Node(T value, Node *prev = nullptr, Node *next = nullptr)
        {
            this->value = value;
            this->prev = prev;
            this->next = next;
        }
        void setNext(Node *next) { this->next = next; }
        void setPrev(Node *prev) { this->prev = prev; }
        void setValue(T value) { this->value = value; }
        Node *getNext() { return this->next; }
        Node *getPrev() { return this->prev; }
        T getValue() { return this->value; }
    };

    Node *header;
    Node *tailer;
    int size;

public:
    DoublyLinkedList()
    {
        this->header = new Node(0);
        this->tailer = new Node(0);

        this->header->setNext(this->tailer);
        this->tailer->setPrev(this->header);

        this->size = 0;
    }

    Node* getHeader() {return this->header;}
    Node* getTailer() {return this->tailer;}
    int getSize() {return this->size;}
    
    void setSize(int newSize) {this->size=newSize;}
    void increaseSize() {this->size++;}
    void decreaseSize() {this->size--;}

    bool isEmpty() {return this->size==0;}

    T first()
    {
        if(this->isEmpty())
            throw "List is empty";
        return this->header->getNext()->getValue(); 
    }
    T last() 
    {
        if(this->isEmpty())
            throw "List is empty";
        return this->tailer->getPrev()->getValue(); 
    }

    void add(T e, Node* p, Node* s)
    {
        Node* newNode = new Node(e, p, s);
        p->setNext(newNode);
        s->setPrev(newNode);
        increaseSize();
    }
    T remove(Node* q)
    {
        q->getPrev()->setNext(q->getNext());
        q->getNext()->setPrev(q->getPrev());
        T temp = q->getValue();
        decreaseSize();
        delete q;
        return temp;
    }

    void addFirst(T value)
    {
        add(value, this->header, this->header->getNext());
    }
    void addLast(T value) {add(value, this->tailer->getPrev(), this->tailer);}

    T removeFirst()
    {   
        if(isEmpty()) throw "List is empty";
        return remove(this->header->getNext());
    }

    T removeLast()
    {   
        if(isEmpty()) throw "List is empty";
        return remove(this->tailer->getPrev());
    }

    T* traverse()
    {
        Node* temp = this->header->getNext();

        T* trav = new T[this->size];
        int i=0;

        while(temp != this->tailer)
        {
            trav[i++] = temp->getValue();   
            temp = temp->getNext();
        }

        return trav;
    }

    T* backTraverse()
    {
        Node* temp = this->tailer->getPrev();

        T* trav = new T[this->size];
        int i=0;

        while(temp != this->header)
        {
            trav[i++] = temp->getValue();   
            temp = temp->getPrev();
        }

        return trav;
    }

};

File: LinkedLists/DoublyLinkedList.java
public class DoublyLinkedList<T> {

    private Node header;
    private Node trailer;
    private int size;

    public DoublyLinkedList() {
        this.header = new Node(null, null, null);
        this.trailer = new Node(null, header, null);
        header.setNext(trailer);
        this.size = 0;
    }

    /* ------------------------ LinkedList Methods ------------------------ */

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public T first() {
        if (isEmpty()) throw new IllegalStateException("LinkedList is empty");
        return header.getNext().getValue();
    }

    public T last() {
        if (isEmpty()) throw new IllegalStateException("LinkedList is empty");
        return trailer.getPrev().getValue();
    }

    private void addBetween(T value, Node before, Node after) {
        Node newNode = new Node(value, before, after);
        before.setNext(newNode);
        after.setPrev(newNode);
        size++;
    }

    public void addFirst(T value) {
        addBetween(value, header, header.getNext());
    }

    public void addLast(T value) {
        addBetween(value, trailer.getPrev(), trailer);
    }

    private T remove(Node node) {
        Node before = node.getPrev();
        Node after = node.getNext();
        before.setNext(after);
        after.setPrev(before);
        size--;
        return node.getValue();
    }

    public T removeFirst() {
        if (isEmpty()) throw new IllegalStateException("LinkedList is empty");
        return remove(header.getNext());
    }

    public T removeLast() {
        if (isEmpty()) throw new IllegalStateException("LinkedList is empty");
        return remove(trailer.getPrev());
    }

    public void traverse() {
        Node node = header.getNext();
        while (node != trailer) {
            System.out.println(node.getValue());
            node = node.getNext();
        }
    }

    /* ------------------------ Getter/Setter ------------------------ */

    public Node getHeader() {
        return header;
    }

    public void setHeader(Node header) {
        this.header = header;
    }

    public Node getTrailer() {
        return trailer;
    }

    public void setTrailer(Node trailer) {
        this.trailer = trailer;
    }

    public int getSize() {
        return size;
    }

    public void setSize(int size) {
        this.size = size;
    }

    /* ------------------------ Node Class ------------------------ */

    class Node {
        private final T value;
        private Node prev;
        private Node next;

        public Node(T v, Node p, Node n) {
            value = v;
            prev = p;
            next = n;
        }

        public T getValue() {
            return value;
        }

        public Node getPrev() {
            return prev;
        }

        public void setPrev(Node prev) {
            this.prev = prev;
        }

        public Node getNext() {
            return next;
        }

        public void setNext(Node next) {
            this.next = next;
        }
    }
}
